* DO NOT MERGE * CORE-17768 Topology changes - Thread per consumer

mbrkic-r3 opened the pull request:
This is alternative solution to https://github.com/corda/corda-runtime-os/pull/4931

Using new topology Kafka topics (flow.start, flow.session, flow.mapper.start, flow.mapper.session.in, flow.mapper.session.out).

Each consumer uses a long running thread to poll messages and store them in an internal buffer. Messages are polled until `maxPollRecords` are polled, then polling is paused until there is again buffer space available. When event mediator calls `poll` it immediately (no poll timeout) gets back messages from consumer's buffer.

Note that config values are hardcoded in [MessageBusConsumer.kt](https://github.com/corda/corda-runtime-os/pull/4878/files#diff-8ce13cdd8a558def0f533ec3ebf729fa4344c7acc93a857b6ea19331a698f387):
```
 private val DEFAULT_POLL_TIMEOUT = Duration.ofMillis(500)
 private const val DEFAULT_MAX_POLL_RECORDS = 500
 private const val BUFFER_QUEUE_TIMEOUT_MILLIS = 100L
```

mbrkic-r3 committed:
CORE-16203 Replace State and Event pattern with Multi-Source Event Mediator in FlowWorker (#4832)

Using Multi-Source Event Mediator instead of State and Event Subscription in FlowWorker.

Polling consumers and committing offsets performed in mediator thread in a synchronous way (without using coroutines)

mbrkic-r3 committed:
CORE-17429 State metadata support in message processor (#4828)

Adds support for state metadata to the state and event processor.

Components using the event mediator to process events and states may wish to set metadata alongside the state that is written back to the underlying state storage. This change adds to the StateAndEventProcessor API to facilitate this, as well as migrating existing state and event processors to use this.

---------

Co-authored-by: James Higgs <james.higgs@r3.com>

mbrkic-r3 committed:
CORE-16203 Removed coroutine usage from Multi-Source Event Mediator.

mbrkic-r3 committed:
CORE-16203 Set linger.ms to 0 for stateAndEvent producer in kafka-messaging-defaults.conf

mbrkic-r3 committed:
CORE-17487 5.1 Performance integration - parallel processing (#4873)

Add support for parallel flows executing within the flow pipeline.

This includes:
- Temporary multi-thread implementation of the state and event pattern.
- Using the SameThreadExecutor for flow fibers, which causes the fiber to run on the thread that started the fiber.
- Using a Kryo Pool for Kryo instances, as kryo instances are not thread-safe themselves.

Add TaskManager and a TaskManagerFactory which is a wrapper around a ScheduledExecutorService used for gathering metrics on executing tasks. The built-in Micrometer metrics for executor services are used, along with some additional metrics in the TaskManager that are more specific to our use-cases.
---------

Co-authored-by: Conal Smith <conal.smith@r3.com>
Co-authored-by: Dan Newton <dan.newton@r3.com>
Co-authored-by: Dries Samyn <dries.samyn@r3.com>

mbrkic-r3 committed:
CORE-17661 5.1 Performance integration - Multi-Source Event Mediator integration with StateManager (#4874)

Integration of Multi-Source Event Mediator with StateManager.

Commit:
Merge pull request #4872 from corda/mbrkic-r3/feature/51-perf-integration

CORE-15419 5.1 Performace Integration

mbrkic-r3 committed:
CORE-16203 5.1 Performance integration - FlowMapper using Multi-Source Event Mediator (#4875)

Using Multi-Source Event Mediator instead of State and Event Subscription in FlowMapperWorker.

---------

Co-authored-by: James Higgs <james.higgs@r3.com>

Omar-awad committed:
Merge pull request #4876 from corda/mbrkic-r3/feature/51-perf-integration-2

CORE-15419 5.1 Performance Integration #2

mbrkic-r3 committed:
CORE-17768 Using new topology Kafka topics (flow.start, flow.session, flow.mapper.start, flow.mapper.session.in, flow.mapper.session.out).

mbrkic-r3 committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17768/topoligy-new-topics

mbrkic-r3 committed:
CORE-17562 Metrics for Multi-Source Event Mediator

mbrkic-r3 committed:
CORE-17562 Fixed unit test

Commit:
Merge branch 'release/os/5.1' into feature/51-perf-integration

mbrkic-r3 commented:
E2E cluster tests are failing due to topic permissions. This should be fine after RPC stuff is integrated (since events should be sent to RPC endpoint instead of those topics).

Commit:
Update API version to fix build (#4882)

JamesHR3 reviewed (commented):
This PR looks reasonable, but I think it needs testing against a cluster with multiple flow and flow mapper workers running. I have a feeling that the flow fiber cache in particular will cause some problems when we have multiple topics in place for a single flow, and we'll need to think about how that works.

Omar-awad committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17562/mediator-metrics-2

Omar-awad committed:
Merge pull request #4879 from corda/mbrkic-r3/CORE-17562/mediator-metrics-2

CORE-17562 Metrics for Multi-Source Event Mediator

thiagoviana committed:
CORE-16242 - Synchronous RPC Pattern API Implementation for Crypto Worker (#4820)

HTTP RPC server integration into the Crypto processor. Allows Crypto processor to receive HTTP external event requests sent from the flow engine. Client integration will come later.

davidcurrie committed:
CORE-17822 Apply prefix to topic overrides (#4893)

JamesHR3 committed:
CORE-17627: Add metadata to flow mapper states for mapper status (#4886)

Adds the state status key to flow mapper state metadata and populates it with the current state.

The metadata key is used by the mapper cleanup logic to ensure that mapper states are removed.

mbrkic-r3 committed:
CORE-17843 Fixed stopping event mediator. Modified  to work with event mediator for flow event processing.

JamesHR3 committed:
CORE-17388: Add session timeout metadata to store alongside the checkpoint (#4889)

Adds session timeout to the metadata stored alongside the checkpoint.

Session timeout is implemented using the state storage lookup mechanism, which requires the metadata to be populated with the session timeout key. The earliest timestamp of session expiry for currently open sessions is the value assigned to this property.

ben-millar committed:
CORE-16181 Implementing RPC client, routing external events through RPC (#4885)

This PR adds a new implementation of MessagingClient, RPCClient, handling synchronous calls to the Corda workers via RPC/HTTP with configurable retries.

Includes also a small bug-fix for the task manager.

---------

Co-authored-by: Miljenko Brkic <97448832+mbrkic-r3@users.noreply.github.com>

mbrkic-r3 committed:
CORE-16181 Fixed integration test

mbrkic-r3 committed:
CORE-16181 Fixed integration test

mbrkic-r3 committed:
CORE-16181 Fixed integration test

mbrkic-r3 committed:
CORE-16181 Fixed integration test

mbrkic-r3 committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17768/topoligy-new-topics

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/CORE-17768/topoligy-new-topics' into mbrkic-r3/CORE-17768/topoligy-new-topics

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/feature/51-perf-integration' into mbrkic-r3/CORE-17768/topoligy-new-topics

# Conflicts:
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt

mbrkic-r3 committed:
CORE-17768 Optimized imports after merge

mbrkic-r3 committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix

JamesHR3 committed:
CORE-17626: Add integration test for flow mapper cleanup (#4907)

Adds an integration test for mapper cleanup.

Also addresses a small bug where the key for the cleanup event was not serialized correctly.

JamesHR3 committed:
CORE-17882: Add a route for flow events (#4915)

Adds a route for flow events to the flow worker event mediator.

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/feature/51-perf-integration' into mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix

# Conflicts:
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/TestStateManagerFactoryImpl.kt

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix' into mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix

Commit:
Merge branch 'release/os/5.1' into feature/51-perf-integration

mbrkic-r3 committed:
Merge branch 'feature/51-perf-integration' into mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix

mbrkic-r3 committed:
CORE-17843 Removed duplicate variable in TestStateManagerFactoryImpl introduced after merge

mbrkic-r3 committed:
CORE-17768 Disabled smoke test "cluster configuration changes are picked up and workers continue to operate normally"

mbrkic-r3 committed:
CORE-17843 Disabled smoke test "cluster configuration changes are picked up and workers continue to operate normally"

mbrkic-r3 committed:
Revert "CORE-17768 Disabled smoke test "cluster configuration changes are picked up and workers continue to operate normally""

This reverts commit eb0945a51033970018555f1045e497fb381704ef.

mbrkic-r3 committed:
CORE-17843 Minor fixed based on review feedback.

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/feature/51-perf-integration' into mbrkic-r3/CORE-17768/topoligy-new-topics

# Conflicts:
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/TestStateManagerFactoryImpl.kt

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/CORE-17843/FlowMapperServiceIntegrationTest-fix-backup' into mbrkic-r3/CORE-17768/topoligy-new-topics

# Conflicts:
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/FlowMapperServiceIntegrationTest.kt
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/TestFlowEventMediatorFactoryImpl.kt
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImpl.kt

mbrkic-r3 committed:
CORE-17768 Fixed compile errors after merge

mbrkic-r3 committed:
CORE-17768 Fixed bad merge

mbrkic-r3 committed:
CORE-17768 Set API version to 5.1.0.37-alpha-1697718230127

mbrkic-r3 committed:
CORE-17768 Set API version to 5.1.0.37-alpha-1697725680919

mbrkic-r3 committed:
CORE-17768 Added logging for debugging purposes

mbrkic-r3 committed:
Revert "CORE-17768 Added logging for debugging purposes"

This reverts commit d78533b57233f9e65fda109df33d88166e7887a4.

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.1' into mbrkic-r3/CORE-17768/topoligy-new-topics

# Conflicts:
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/FlowMapperServiceIntegrationTest.kt
#	components/flow/flow-mapper-service/src/integrationTest/kotlin/net/corda/session/mapper/service/integration/TestFlowEventMediatorFactoryImpl.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/messaging/FlowEventMediatorFactoryImplTest.kt

mbrkic-r3 committed:
CORE-17768 Routing FlowEvents that FlowEngine sends out to itself to "flow.event" topic

mbrkic-r3 committed:
CORE-17768 Fixed failing FlowEventMediatorFactoryImplTest

mbrkic-r3 committed:
CORE-17768 Using long-running thread per consumer

mbrkic-r3 committed:
CORE-17768 Fixed unit tests

mbrkic-r3 committed:
CORE-17768 Fixed unit tests

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.1' into mbrkic-r3/CORE-17768/topoligy-new-topics

# Conflicts:
#	.run/Combined Worker Local.run.xml
#	libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/MultiSourceEventMediatorImpl.kt

mbrkic-r3 committed:
CORE-17768 Using beta version of API
